File: src/10.graph/42.547.ts
export function findCircleNum1(isConnected: number[][]): number {
    const n = isConnected.length;
    const used = new Array(n).fill(false);
    const dfs = (i: number) => {
        used[i] = true;
        for (let j = 0; j < n; j++) {
            if (isConnected[i][j] === 1 && !used[j]) {
                dfs(j);
            }
        }
    };
    let k = 0;

    for (let i = 0; i < n; i++) {
        if (!used[i]) {
            k++;
            dfs(i);
        }
    }

    return k;
}

export function findCircleNum(isConnected: number[][]): number {
    const n = isConnected.length;
    const uf = new UnionFind();

    for (let i = 0; i < n; i++) {
        uf.add(i);

        // 取矩阵左下方的数据
        for (let j = 0; j < i; j++) {
            if (isConnected[i][j] === 1) {
                uf.merge(i, j);
            }
        }
    }

    return uf.numOfSets;
}

class UnionFind {
    father: Map<number, number>;
    numOfSets: number;
    constructor() {
        this.father = new Map<number, number>();
        this.numOfSets = 0;
    }

    add(x: number) {
        if (!this.father.has(x)) {
            this.father.set(x, null);
            this.numOfSets++;
        }
    }

    find(x: number) {
        let root = x;

        while (this.father.get(root) !== null) {
            root = this.father.get(root);
        }

        while (x !== root) {
            let originFather = this.father.get(x);
            this.father.set(x, root);
            x = originFather;
        }

        return root;
    }

    merge(x: number, y: number) {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX !== rootY) {
            this.father.set(rootX, rootY);
            this.numOfSets--;
        }
    }

    isConnected(x: number, y: number) {
        return this.find(x) === this.find(y);
    }
}

File: src/10.graph/42.547.go
package graph

func FindCircleNum1(isConnected [][]int) int {
	n := len(isConnected)
	used := make([]bool, n)
	var dfs func(i int)
	dfs = func(i int) {
		used[i] = true
		for j := 0; j < n; j++ {
			if isConnected[i][j] == 1 && !used[j] {
				dfs(j)
			}
		}
	}

	count := 0

	for i := 0; i < n; i++ {
		if !used[i] {
			count++
			dfs(i)
		}
	}

	return count
}

func FindCircleNum(isConnected [][]int) int {
	n := len(isConnected)
	uf := &UnionFind{}

	uf.init(n)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if isConnected[i][j] == 1 {
				uf.union(i, j)
			}
		}
	}

	return uf.size
}

type UnionFind struct {
	roots []int
	size  int
}

func (u *UnionFind) init(n int) {
	u.roots = make([]int, n)
	for i := 0; i < n; i++ {
		u.roots[i] = i
	}
	u.size = n
}

func (u *UnionFind) find(i int) int {
	if i == u.roots[i] {
		return i
	}

	u.roots[i] = u.find(u.roots[i])

	return u.roots[i]
}

func (u *UnionFind) union(x int, y int) {
	rootX := u.find(x)
	rootY := u.find(y)

	if rootX != rootY {
		u.roots[rootX] = rootY
		u.size--
	}
}
